app/validation/validator.py

```diff
@@ -347,39 +352,77 @@ def validate_skip_condition(self, skip_condition, answer_ids_with_group_id, bloc
 
-                if not self.options_contain_value(answer['options'], add_answer_value):
+                if not self.options_contain_value(collector_answer['options'], add_answer_value):
                     errors.append(
                         self._error_message('The list collector block {} has an add_answer_value that is not present in the answer values'
                                             .format(block['id'])))
 
-        if 'routing_rules' in block['add_block']:
-            errors.append(self._error_message('The list collector block {} contains routing rule on the "add_block"'
-                                              .format(block['id'])))
+        for remove_question in remove_questions:
+            for remove_answer in remove_question['answers']:
+                if remove_answer['type'] != 'Radio':
+                    errors.append(self._error_message('The list collector remove block {} does not contain a Radio answer type'
+                                                      .format(block['id'])))
 
-        if 'routing_rules' in block['edit_block']:
-            errors.append(self._error_message('The list collector block {} contains routing rule on the "edit_block"'
-                                              .format(block['id'])))
+                if not self.options_contain_value(remove_answer['options'], remove_answer_value):
+                    errors.append(
+                        self._error_message('The list collector block {} has a remove_answer_value that is not present in the answer values'
+                                            .format(block['id'])))
+
+        nested_blocks = [('add_block', 'ListAddQuestion'), ('edit_block', 'ListEditQuestion'), ('remove_block', 'ListRemoveQuestion')]
+        for nested_block_name, nested_block_type in nested_blocks:
+            nested_block = block[nested_block_name]
+            if nested_block['type'] != nested_block_type:
+                errors.append(self._error_message(
+                    f'The type of the {nested_block_name} is incorrect for a nested ListCollector block. '
+                    f'Expected: {nested_block_type}'))
+            if 'routing_rules' in nested_block:
+                errors.append(self._error_message(f'The list collector block {block["id"]} contains routing rules '
+                                                  f'on the {nested_block["id"]} sub block'))
+
+        errors.extend(self.validate_list_collector_answer_ids(block))
+
+        return errors
+
+    def validate_list_collector_answer_ids(self, block):
+        """
+        - Ensure that answer_ids on add blocks match between all blocks that populate a single list.
+        - Enforce the same answer_ids on add and edit sub-blocks
+        """
+        errors = []
+        list_name = block['populates_list']
+
+        add_block_questions = self._get_all_questions_for_block(block['add_block'])
```

Comment:
All references by key that is not using `.get()` isn't validating the schema since it fails on key error. It is mandatory in the schemas but it doesn't tell you that until it does our python check also.

```python
File "/Users/abrahm/Documents/projects/eq-schema-validator/app/validation/validator.py", line 358, in validate_list_collector_type
    remove_questions = self._get_all_questions_for_block(block['remove_block'])
KeyError: 'remove_block'
```

Reply:
Still an issue.